src/keyword_vpd_parser.cpp

```diff
@@ -0,0 +1,148 @@
+#include "keyword_vpd_parser.hpp"
+#include "constants.hpp"
+
+#include <iostream>
+#include <numeric>
+#include <string>
+
+namespace vpd
+{
+namespace keyword
+{
+
+/* @brief Keyword VPD parser function*/
+std::variant<types::KeywordVpdMap, types::ParsedVPD> KeywordVpd::parse()
+{
+    vpdIterator = vpdVector.begin();
+
+    checkNextBytesValidity(sizeof(constants::KW_VPD_START_TAG));
+    std::advance(vpdIterator, sizeof(constants::KW_VPD_START_TAG));
+
+    int isBonoVpd = checkIsBonoVpd();
+    auto kwValMap = parser();
+
+    // Do these validations before returning parsed data.
+    // Skip these validation if it is bono type VPD
+    if (!isBonoVpd)
+    {
+        validateKwVpdDataEndTag();
+        validateChecksum();
+    }
+    validateKwVpdEndTag();
+
+    return kwValMap;
+}
+
+int KeywordVpd::checkIsBonoVpd()
+{
+    size_t dataSize = getKwDataSize();
+
+    checkNextBytesValidity(constants::TWO_BYTES + dataSize);
+
+    // +TWO_BYTES is the description's size byte
+    std::advance(vpdIterator, constants::TWO_BYTES + dataSize);
+
+    int kwVpdType = 0;
+    // Check for invalid vendor defined large resource type
+    if (*vpdIterator != constants::KW_VPD_START_TAG)
+    {
+        if (*vpdIterator != constants::ALT_KW_VPD_START_TAG)
+        {
+            throw std::runtime_error("Invalid Keyword Vpd Start Tag");
```

Comment:
Why runtime error and not VPD data exception?

Reply:
Ack.